data_diff/database.py

```diff
@@ -726,63 +787,35 @@ def connect_to_uri(db_uri: str, thread_count: Optional[int] = 1) -> Database:
 
     Note: For non-cloud databases, a low thread-pool size may be a performance bottleneck.
 
-    Supported databases:
+    Supported schemes:
     - postgres
     - mysql
     - mssql
     - oracle
     - snowflake
     - bigquery
     - redshift
+    - presto
     """
 
     dsn = dsnparse.parse(db_uri)
     if len(dsn.schemes) > 1:
         raise NotImplementedError("No support for multiple schemes")
     (scheme,) = dsn.schemes
 
-    if scheme == "snowflake":
-        if len(dsn.paths) == 1:
-            (database,) = dsn.paths
-            schema = dsn.query["schema"]
-        elif len(dsn.paths) == 2:
-            database, schema = dsn.paths
-        else:
-            raise ValueError(f"Too many parts in path. Expected format: '{HELP_SNOWFLAKE_URI_FORMAT}'")
-        try:
-            warehouse = dsn.query["warehouse"]
-        except KeyError:
-            raise ValueError(f"Must provide warehouse. Expected format: '{HELP_SNOWFLAKE_URI_FORMAT}'")
-        return Snowflake(dsn.host, dsn.user, dsn.password, warehouse=warehouse, database=database, schema=schema)
-    elif scheme == "presto":
-        if len(dsn.paths) == 1:
-            (catalog,) = dsn.paths
-            schema = dsn.query.get("schema")
-        elif len(dsn.paths) == 2:
-            catalog, schema = dsn.paths
-        else:
-            raise ValueError(f"Too many parts in path. Expected format: '{HELP_PRESTO_URI_FORMAT}'")
-
-        return Presto(dsn.host, dsn.port, dsn.user, dsn.password, catalog=catalog, schema=schema)
-
-    if len(dsn.paths) == 0:
-        path = ""
-    elif len(dsn.paths) == 1:
-        (path,) = dsn.paths
-    else:
-        raise ValueError("Bad value for uri, too many paths: %s" % db_uri)
-
-    if scheme == "postgres":
-        return Postgres(dsn.host, dsn.port, path, dsn.user, dsn.password, thread_count=thread_count)
-    elif scheme == "mysql":
-        return MySQL(dsn.host, dsn.port, path, dsn.user, dsn.password, thread_count=thread_count)
-    elif scheme == "mssql":
-        return MsSQL(dsn.host, dsn.port, path, dsn.user, dsn.password, thread_count=thread_count)
-    elif scheme == "bigquery":
-        return BigQuery(dsn.host, path)
-    elif scheme == "redshift":
-        return Redshift(dsn.host, dsn.port, path, dsn.user, dsn.password, thread_count=thread_count)
-    elif scheme == "oracle":
-        return Oracle(dsn.host, dsn.port, path, dsn.user, dsn.password, thread_count=thread_count)
-
-    raise NotImplementedError(f"Scheme {dsn.scheme} currently not supported")
+    try:
+        matcher = MATCH_URI_PATH[scheme]
+    except KeyError:
+        raise NotImplementedError(f"Scheme {scheme} currently not supported")
+
+    cls = matcher.database_cls
+    kw = matcher.match_path(dsn)
+
+    if scheme == "bigquery":
+        return cls(dsn.host, **kw)
+
+    if issubclass(cls, ThreadedDatabase):
+        return cls(dsn.host, dsn.port, dsn.user, dsn.password, thread_count=thread_count, **kw)
```

Comment:
Is there a way we can just pass arbitrary key-word arguments? For example, say I want to connect to Postgres with `sslmode='require'`, I would expect to be able to pass that as part of the URI

Reply:
That's what **kw is for.

Reply:
Right, should've been more specific, for e.g. the Postgres driver they're not passed along:

```
postgres://postgres:Password1@localhost/postgres?sslmode=require
```

Will give `TypeError: __init__() got an unexpected keyword argument 'sslmode'`

Reply:
Oh, you mean pass everything directly to the postgres driver.. I guess we can do that. I was worried it would might cause confusion or strange behaviors, though I can't say exactly how or why.

Reply:
We can wait until it arises as an issue if you like. I just think that'd be what I expected :)

Reply:
It's fine, I don't see any harm in it. Added.